src/routes/calendar.rs

```diff
@@ -1,95 +1,113 @@
     })?;
 
+    // no-op if course is already in state
+    if user_state
+        .selection
+        .iter()
+        .any(|c| c.0.subject_code == subject_code && c.0.course_code == course_code)
+    {
+        return Ok((CookieJar::new(), courses_container()));
+    }
+
     // get a db conn
     let conn = state.get_conn(&term).ok_or(StatusCode::NOT_FOUND)?;
 
     // query db
-    let (subject_code, course_code) = conn
-        .prepare("SELECT subject_code, course_code FROM section WHERE crn = ?1;")
+    let sections = conn
+        .prepare(
+            "SELECT sequence_code, crn FROM section WHERE subject_code = ?1 AND course_code = ?2",
+        )
         .context("failed to prepare courses SQL statement")?
-        .query_row([form.crn], |row| {
-            let sub: String = row.get(0)?;
-            let course: String = row.get(1)?;
-            Ok((sub, course))
+        .query_and_then((&subject_code, &course_code), |row| {
+            let sequence_code: String = row.get(0)?;
+            let crn: u64 = row.get(1)?;
+            Ok((sequence_code, crn))
         })
-        .context("query failed, course not found")
-        .map_err(|err| {
-            debug!(?err);
-            StatusCode::NOT_FOUND
-        })?;
+        .context("query failed")?
+        .collect::<anyhow::Result<Vec<_>>>()?;
 
-    // building response
-    let found = user_state.selection.iter().any(|thincourse| {
-        thincourse.subject_code == subject_code && thincourse.course_code == course_code
-    });
+    let mut default_sections = HashMap::new();
+    for (sequence_code, crn) in sections {
+        let (letter, rest) = sequence_code.split_at(1);
+        match default_sections.entry(letter.to_string()) {
+            Entry::Vacant(e) => {
+                e.insert((rest.to_string(), crn));
+            }
+            Entry::Occupied(mut e) => {
+                if *rest < *e.get().0 {
+                    e.insert((rest.to_string(), crn));
+                }
+            }
+        }
+    }
+    let mut default_sections = default_sections.into_iter().collect::<Vec<_>>();
+    default_sections.sort_by_cached_key(|t| t.0.clone());
+    let default_sections = default_sections
+        .into_iter()
+        .map(|t| ThinSection { crn: t.1 .1 })
+        .collect::<Vec<_>>();
 
-    let mut jar = CookieJar::new();
-    if !found {
-        // new course, new cookie
-        let mut user_state = user_state.to_owned();
-        user_state.selection.push(ThinCourse {
-            subject_code,
-            course_code,
-            sections: Vec::new(),
-        });
+    debug!(?default_sections);
 
-        let cookie: Cookie = user_state.into();
-        jar = jar.add(cookie);
-    }
+    let jar = CookieJar::new().add({
+        let mut user_state = user_state.clone();
+        user_state.selection.push((
+            ThinCourse {
+                subject_code,
+                course_code,
+            },
+            default_sections,
```

Comment:
I wasn't sure what to do with this Vec, so I decided to prepopulate it here.

Imo there are two options for when the user hasn't selected any sections yet:

1. keep it empty internally and render as selecting the first one, or
2. store that they've selected the first one and don't adjust the rendering logic.

It made more sense for me to do the 2nd one, since it reduces the work to only once on mutation. I'm not sure if your intention was to eventually do the 1st one @hn275, so let me know if I'm misunderstanding something here.

Reply:
see https://github.com/brennanmcmicking/scheduler/pull/26#discussion_r1708720742